### PR TITLE
Fix pinning expressions for prerelease builds

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
cc @mbargull

**Comment by mbargull** [comment]
Ah, didn't see this before, thanks for the cc!
I think we have multiple issues here:
1. `apply_pin_expressions` does not respect certain corner cases like pre-releases, epochs and thus needs fixing
2. `apply_pin_expressions` should not be run on build string-pinned deps (because you also get things like `python 3.10.0 h1234567` => `python >=3.10,<3.11.0a0` which is not correct; see `conda search --info conda-forge::libpython-static`)

For 2.: Maybe something along this (in `conda_build.render`)
```diff
 def get_pin_from_build(m, dep, build_dep_versions):
+    match_spec = MatchSpec(dep)  # from conda.exports import MatchSpec
+    if match_spec.get_exact_value('version') and match_spec.get_exact_value('build'):
+        # exact pin; don't apply pin expressions
+        return dep 
     dep_split = dep.split()
```
might work. I haven't tried this one yet, though.

**Comment by isuruf** [comment]
If 1 is fixed (at least the pre-release part) as in this PR, then 2 becomes not so important. 2 would be nice to have, but not necessary. Note that fixing 2 alone is not enough for our python issue if we start building downstream packages with rc0.

**Comment by mbargull** [comment]
> Note that fixing 2 alone is not enough for our python issue if we start building downstream packages with rc0.

Yep, I know. Either we fix 1 and get a new release with that fix or we'd have to mark the RC as `python=3.10.0=rc_h123457` or similar as a workaround (and then of course always have it remain in the pre-release channel, i.e., never `main` label).

**Comment by isuruf** [comment]
> Either we fix 1 and get a new release with that fix 

Which is what this PR does

File: conda_build/utils.py [hunk 1]
```diff
@@ -1514,7 +1514,12 @@ def apply_pin_expressions(version, min_pin='x.x.x.x.x.x.x', max_pin='x'):
             if versions[p_idx][-1] == '.':
                 versions[p_idx] = versions[p_idx][:-1]
     if versions[0]:
-        versions[0] = '>=' + versions[0]
+        if VersionOrder(version) < VersionOrder(versions[0]):
```
Comment (mbargull) [inline conda_build/utils.py:1517]: Maybe
```suggestion
        if (
            VersionOrder(version) < VersionOrder(versions[0])
            and not MatchSpec('pkg ' + version).match(MatchSpec('pkg ' + versions[0]))
        ):
```
could work to handle the `version = '1.2.*'` case?

**Comment by mbargull** [comment]
> > Either we fix 1 and get a new release with that fix
> 
> Which is what this PR does

yeah, that was mostly meant regarding "and get a new release" ;).

File: conda_build/utils.py (hunk 1 above)
Comment (isuruf) [reply conda_build/utils.py:1517]: That doesn't handle the case `1.2.3.*`

**Comment by isuruf** [comment]
This is ready for a review

**Comment by mbargull** [comment]
Merging this since all CI failures are unrelated to the PR.

@chenghlee, @jezdez, if we want to go ahead with a RC-based migration on conda-forge ( https://github.com/conda-forge/python-feedstock/pull/481 , https://github.com/conda-forge/conda-forge.github.io/issues/1499 ), we'd need a new `conda-build` release for this. I'm not aware of any other things we'd need though; @isuruf might know.

**Comment by isuruf** [comment]
Thanks @mbargull. I need https://github.com/conda/conda-build/pull/4248 as well for building GCC.

**Pull request merged**